src/raftstore/store/fsm/store.rs

```diff
@@ -1325,35 +1337,56 @@ impl<'a, T: Transport, C: PdClient> StoreFsmDelegate<'a, T, C> {
             return Ok(false);
         }
 
-        let start_key = data_key(msg.get_start_key());
-        if let Some((_, exist_region_id)) = meta
-            .region_ranges
-            .range((Excluded(start_key), Unbounded::<Key>))
-            .next()
-        {
-            let exist_region = &meta.regions[&exist_region_id];
-            if enc_start_key(exist_region) < data_end_key(msg.get_end_key()) {
-                debug!("msg {:?} is overlapped with region {:?}", msg, exist_region);
-                if util::is_first_vote_msg(msg.get_message()) {
-                    meta.pending_votes.push(msg.to_owned());
-                }
-                self.ctx.raft_metrics.message_dropped.region_overlap += 1;
-
-                // Make sure the range of region from msg is covered by existing regions.
-                // If so, means that the region may be generated by some kinds of split
-                // and merge by catching logs. So there is no need to accept a snapshot.
-                if !is_range_covered(
-                    &meta.region_ranges,
-                    |id: u64| &meta.regions[&id],
-                    data_key(msg.get_start_key()),
-                    data_end_key(msg.get_end_key()),
+        let mut regions_to_destroy = vec![];
+        for (_, id) in meta.region_ranges.range((
+            Excluded(data_key(msg.get_start_key())),
+            Unbounded::<Vec<u8>>,
+        )) {
+            let exist_region = &meta.regions[&id];
+            if enc_start_key(exist_region) >= data_end_key(msg.get_end_key()) {
+                break;
+            }
+
+            debug!("msg {:?} is overlapped with region {:?}", msg, exist_region);
+            if util::is_first_vote_msg(msg.get_message()) {
+                meta.pending_votes.push(msg.to_owned());
+            }
+
+            // Make sure the range of region from msg is covered by existing regions.
+            // If so, means that the region may be generated by some kinds of split
+            // and merge by catching logs. So there is no need to accept a snapshot.
+            if !is_range_covered(
+                &meta.region_ranges,
+                |id: u64| &meta.regions[&id],
+                data_key(msg.get_start_key()),
+                data_end_key(msg.get_end_key()),
+            ) {
+                if maybe_destroy_source(
+                    meta,
+                    region_id,
+                    exist_region.get_id(),
+                    msg.get_region_epoch().to_owned(),
                 ) {
-                    meta.pending_cross_snap
-                        .insert(region_id, msg.get_region_epoch().to_owned());
+                    regions_to_destroy.push(exist_region.get_id());
```

Comment:
Why not send `MergeResult` here?

Reply:
there may be several overlapping regions, and other regions may not pass the check. So we only destroy them only if all the overlapping regions pass the check.